opencog/cogserver/server/AgentRunnerThread.cc

```diff
@@ -0,0 +1,167 @@
+ * along with this program; if not, write to:
+ * Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include <opencog/cogserver/server/AgentRunnerThread.h>
+#include <opencog/util/Logger.h>
+
+using namespace std;
+
+
+namespace opencog
+{
+
+AgentRunnerThread::AgentRunnerThread(const std::string &name) :
+        AgentRunnerBase(name), runAgents(false), agentsModified(false),
+        clearAll(false)
+{
+}
+
+AgentRunnerThread::~AgentRunnerThread()
+{
+    removeAllAgents();
+    joinRunThread();
+}
+
+void AgentRunnerThread::start()
+{
+    {
+        // locking is required, to prevent dead-lock in condition_variable wait()
+        lock_guard<mutex> lock(runAgentsMutex);
+        runAgents.store(true, memory_order_relaxed);
+    }
+    runAgentsCond.notify_one();
+}
+
+void AgentRunnerThread::stop()
+{
+    runAgents.store(false, memory_order_relaxed);
+}
+
+void AgentRunnerThread::addAgent(AgentPtr a)
+{
+    if (!hasAgents()) {
+        joinRunThread();
+        /*
+         * processAgentsThread() will exit immediately if no agents are available,
+         * so we add the first one immediately
+         */
+        {
+            lock_guard<mutex> lock(agentsMutex);
+            AgentRunnerBase::addAgent(a);
+        }
+        runThread = thread(&AgentRunnerThread::processAgentsThread, this);
+        // todo: set thread name when using pthreads
+    }
+    else {
+        lock_guard<mutex> lock(agentsMutex);
+        agentsAddQ.push_back(a);
+        agentsModified.store(true, memory_order_relaxed);
+    }
+}
+
+void AgentRunnerThread::removeAgent(AgentPtr a)
+{
+    lock_guard<mutex> lock(agentsMutex);
+    agentsRemoveQ.push_back(a);
+    agentsModified.store(true, memory_order_relaxed);
+}
+
+void AgentRunnerThread::removeAllAgents(const std::string& id)
+{
+    lock_guard<mutex> lock(agentsMutex);
+    idsRemoveQ.push_back(id);
+    agentsModified.store(true, memory_order_relaxed);
+}
+
+void AgentRunnerThread::removeAllAgents()
+{
+    lock_guard<mutex> lock(agentsMutex);
+    clearAll = true;
+    agentsModified.store(true, memory_order_relaxed);
+}
+
+const AgentSeq& AgentRunnerThread::getAgents() const
+{
+    lock_guard<mutex> lock(agentsMutex);
+    return agents;
+}
+
+bool AgentRunnerThread::hasAgents() const
+{
+    lock_guard<mutex> lock(agentsMutex);
+    return !agents.empty();
+}
+
+void AgentRunnerThread::processAgentsThread()
+{
+    /*
+     * 'agents' is only modified in this function, so there is no need to
+     * protect read accesses to this container in it.
+     */
+    logger().debug("[CogServer::%s] Agent thread started", name.c_str());
+    while (!agents.empty()) {
+        if (!runAgents.load(memory_order_relaxed)) {
+            unique_lock<mutex> lock(runAgentsMutex);
+            runAgentsCond.wait(lock,
+                [this] {return runAgents.load(memory_order_relaxed);});
```

Comment:
I didn't want to protect agents.empty(), as I don't like to lock a mutex each time while condition is being checked. As nobody else is modifying 'agents', I think it is safe to read it without protection in this function although it is not atomic. Can you explain why agents need protection while no other method will change it when this code is being run?